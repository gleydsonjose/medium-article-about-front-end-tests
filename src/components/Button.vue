<template>
  <button
    class="button"
    :style="buttonStyle"
    :disabled="disabled"
    @click="clickEvent"
  >
    <i v-show="useIcon" class="mdi button__icon" :class="colorsType.icon"></i>
    <p class="button__text">{{ text }}</p>
  </button>
</template>

<script>
import { COLORS_TYPE } from "@/constants/colors.js"

export default {
  name: "Button",
  props: {
    disabled: {
      type: Boolean,
      default: false,
    },
    useIcon: {
      type: Boolean,
      default: false,
    },
    text: {
      type: String,
      default: "",
    },
    type: {
      type: String,
      default: "default",
      validator: (type) =>
        ["default", "info", "success", "warn", "error"].includes(type),
    },
  },
  computed: {
    colorsType() {
      return COLORS_TYPE[this.type];
    },
    buttonStyle() {
      const style = {
        "background-color": this.colorsType.backgroundColor,
        "border-color": this.colorsType.borderColor,
      };

      return style;
    },
  },
  methods: {
    clickEvent() {
      this.$emit("click");
    },
  },
};
</script>

<style scoped>
.button {
  display: flex;
  justify-content: center;
  align-items: center;
  border-style: solid;
  border-width: 1px;
  border-radius: 5px;
  color: #eceff1;
  padding: 14px;
  transition: opacity 0.3s ease-out;
  font-size: 12pt;
}

.button:hover {
  opacity: 0.9;
  cursor: pointer;
  transition: opacity 0.3s ease-in;
}

.button__icon {
  margin-right: 5px;
}

.button__text {
  font-weight: bold;
  margin: 0;
}
</style>
