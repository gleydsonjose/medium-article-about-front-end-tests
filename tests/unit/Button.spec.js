import { shallowMount } from "@vue/test-utils";
import Button from "@/components/Button.vue";

describe("<template/>", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("deveria mostrar o icone após usar prop useIcon", () => {
    const wrapper = shallowMount(Button, {
      propsData: { useIcon: true },
    });

    expect(wrapper.find(".button__icon").isVisible()).toBeTruthy();
  });

  it("deveria esconder o icone após não usar prop useIcon", () => {
    const wrapper = shallowMount(Button);

    expect(wrapper.find(".button__icon").isVisible()).toBeFalsy();
  });

  it("deveria renderizar o texto recebido por prop", () => {
    const text = "Clique aqui para salvar";
    const wrapper = shallowMount(Button, {
      propsData: {
        text,
      },
    });

    expect(wrapper.find(".button__text").text()).toBe(text);
  });

  it("deveria acionar um evento após clicar no botão", () => {
    const clickEventSpy = jest.spyOn(Button.methods, "clickEvent");
    const wrapper = shallowMount(Button);

    wrapper.find(".button").trigger("click");

    expect(clickEventSpy).toBeCalled();
  });

  it("não deveria acionar um evento após clicar no botão que está desabilitado", () => {
    const clickEventMethodSpy = jest.spyOn(Button.methods, "clickEvent");
    const wrapper = shallowMount(Button, { propsData: { disabled: true } });

    wrapper.find(".button").trigger("click");

    expect(clickEventMethodSpy).toBeCalledTimes(0);
  });

  it("deveria ser igual ao snapshot", () => {
    const wrapper = shallowMount(Button);

    expect(wrapper).toMatchSnapshot();
  });
});

describe("props:type", () => {
  it("deveria retornar erro caso o tipo recebido pela prop não seja aceitável", () => {
    expect(Button.props.type.validator("danger")).toBeFalsy();
  });
});

describe("methods@clickEvent", () => {
  it("deveria chamar emit após usar método", () => {
    const wrapper = shallowMount(Button);
    wrapper.find(".button").trigger("click");
    expect(wrapper.emitted().click).toBeTruthy();
  });
});
