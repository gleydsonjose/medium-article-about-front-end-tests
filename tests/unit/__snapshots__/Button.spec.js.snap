// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<template/> deveria ser igual ao snapshot 1`] = `
<button class="button" style="background-color: rgb(55, 71, 79); border-color: #263238;"><i class="mdi button__icon mdi-alert-circle" style="display: none;"></i>
  <p class="button__text"></p>
</button>
`;
